mqttclients/c/tests/mqtt_client_test.c

```diff
@@ -0,0 +1,248 @@
+{
+  mqtt_client_test_state* test_state = (mqtt_client_test_state*)state;
+  mqtt_client_connection_settings* connection_settings = test_state->connection_settings;
+
+  setenv("MQTT_TCP_PORT", invalid_env_var, 1);
+  assert_false(set_int_connection_setting(
+      &connection_settings->tcp_port, "MQTT_TCP_PORT", DEFAULT_TCP_PORT));
+  assert_null(connection_settings->tcp_port);
+}
+
+// Test successful setting of an bool environment variable's default value
+static void test_set_bool_connection_setting_default_value_sucess(void** state)
+{
+  mqtt_client_test_state* test_state = (mqtt_client_test_state*)state;
+  mqtt_client_connection_settings* connection_settings = test_state->connection_settings;
+
+  assert_true(set_bool_connection_setting(
+      &connection_settings->clean_session, "MQTT_CLEAN_SESSION", DEFAULT_CLEAN_SESSION));
+  assert_bool_equal(connection_settings->clean_session, DEFAULT_CLEAN_SESSION);
+}
+
+// Test successful setting of an bool environment variable
+static void test_set_bool_connection_setting_sucess(void** state)
+{
+  mqtt_client_test_state* test_state = (mqtt_client_test_state*)state;
+  mqtt_client_connection_settings* connection_settings = test_state->connection_settings;
+
+  setenv("MQTT_CLEAN_SESSION", valid_clean_session_str, 1);
+  assert_true(set_bool_connection_setting(
+      &connection_settings->clean_session, "MQTT_CLEAN_SESSION", DEFAULT_CLEAN_SESSION));
+  assert_bool_equal(connection_settings->clean_session, valid_clean_session);
+}
+
+// Test failure if invalid bool environment variable is defined
+static void test_set_bool_connection_setting_invalid_bool_failure(void** state)
+{
+  mqtt_client_test_state* test_state = (mqtt_client_test_state*)state;
+  mqtt_client_connection_settings* connection_settings = test_state->connection_settings;
+
+  setenv("MQTT_CLEAN_SESSION", invalid_env_var, 1);
+  assert_false(set_bool_connection_setting(
+      &connection_settings->clean_session, "MQTT_CLEAN_SESSION", DEFAULT_CLEAN_SESSION));
+  assert_null(connection_settings->clean_session);
+}
+
+// Test failed setting of all connection settings if no environment variables are defined
+static void test_mqtt_client_set_connection_settings_no_env_vars_failure(void** state)
+{
+  mqtt_client_test_state* test_state = (mqtt_client_test_state*)state;
+  mqtt_client_connection_settings* connection_settings = test_state->connection_settings;
+
+  assert_false(mqtt_client_set_connection_settings(connection_settings));
+}
+
+// Test minimum successful setting of all connection settings
+static void test_mqtt_client_set_connection_settings_min_sucess(void** state)
+{
+  mqtt_client_test_state* test_state = (mqtt_client_test_state*)state;
+  mqtt_client_connection_settings* connection_settings = test_state->connection_settings;
+
+  setenv("MQTT_HOST_NAME", valid_host_name, 1);
+
+  assert_true(mqtt_client_set_connection_settings(connection_settings));
```

Comment:
These are not actually unit-tests since they are using Mosquitto MQTT client as far as I understand.
I would call them "integration tests" instead - they are still necessary.

Unit testing should be reserved for testing of our components exclusively (without any dependencies included).

Reply:
So far these aren't using any Mosquitto client code, but I'll make sure to separate the two concepts once we do!